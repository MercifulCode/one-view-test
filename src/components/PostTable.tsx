import * as React from "react";
import { connect } from "react-redux";
import { TableRow } from "./TableRow";
import { fetchPostData } from "../reducers/JSONReducer";
import { bindActionCreators, Dispatch } from "redux";

export interface IPostTableProps {
  currentUser: IUserResponse;
  data: IPostResponse[];
  dispatchJSONFetch(fetchFn: any): void;
}

export class PostTableClass extends React.Component<IPostTableProps> {
  public componentDidMount() {
    const { currentUser } = this.props;
    this.props.dispatchJSONFetch(
      `https://jsonplaceholder.typicode.com/posts?userId=${currentUser.id}`
    );
  }

  public componentDidUpdate(prevProps: IPostTableProps) {
    const { currentUser, dispatchJSONFetch } = this.props;
    if (currentUser.id !== prevProps.currentUser.id) {
      dispatchJSONFetch(
        `https://jsonplaceholder.typicode.com/posts?userId=${currentUser.id}`
      );
    }
  }

  public render() {
    const { data } = this.props;
    return (
      <table>
        <thead>
          <TableRow cellContent={["Title", "Body"]} />
        </thead>
        <tbody>
          {data.map((datum, index) => (
            <TableRow
              key={`post_table_row_${index}`}
              cellContent={[datum.title, datum.body]}
            />
          ))}
        </tbody>
      </table>
    );
  }
}

const mapStateToProps = (state: { postData: IPostResponse[] }) => ({
  data: state.postData,
});

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      dispatchJSONFetch: fetchPostData,
    },
    dispatch
  );

export const PostTable = connect(
  mapStateToProps,
  mapDispatchToProps
)(PostTableClass);
