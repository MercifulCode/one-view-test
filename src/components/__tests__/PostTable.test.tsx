import { PostTableClass } from "../PostTable";
import { shallow } from "enzyme";

describe("PostTable", () => {
  const aUser: IUserResponse = {
    address: {
      city: "Springfield",
      street: "Fake St",
      suite: "123",
      zipcode: "12345",
      geo: {
        lat: "0",
        lng: "0",
      },
    },
    name: "Homer Simpson",
    company: {
      name: "Nuclear Power Plant",
      catchPhrase: "Excellent",
      bs: "Mmmmmm bees",
    },
    email: "chunkylover53@aol.com",
    id: 1,
    phone: "123-456-7890",
    username: "HomerJay",
    website: "",
  };

  it("Should render when given a user.", () => {
    const wrapper = shallow(
      <PostTableClass
        currentUser={aUser}
        data={[]}
        dispatchJSONFetch={jest.fn()}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
