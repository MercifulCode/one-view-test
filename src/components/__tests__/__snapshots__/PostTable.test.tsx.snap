// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostTable Should render when given a user. 1`] = `
<table>
  <thead>
    <TableRow
      cellContent={
        Array [
          "Title",
          "Body",
        ]
      }
    />
  </thead>
  <tbody />
</table>
`;
